test: add ParsePacketsFromEvents

tbruyelle opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Returns all the packets found in the events. Comes in addition with ParsePacketFromEvents (singular version), which only returns the first packet found in the events.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #4287


### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)


<!--
Example commit messages:

fix: skip emission of unpopulated memo field in ics20
deps: updating sdk to v0.46.4
chore: removed unused variables
e2e: adding e2e upgrade test for ibc-go/v6
docs: ics27 v6 documentation updates
feat: add semantic version utilities for e2e tests
feat(api)!: this is an api breaking feature
fix(statemachine)!: this is a statemachine breaking fix
-->

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

charleenfei reviewed (approved):
thanks @tbruyelle 🙏

srdtrk commented on testing/events.go:
```diff
@@ -120,6 +120,22 @@ func ParsePacketFromEvents(events []abci.Event) (channeltypes.Packet, error) {
+	var packets []channeltypes.Packet
+	for i, ev := range events {
+		if ev.Type == channeltypes.EventTypeSendPacket {
+			packet, err := ParsePacketFromEvents(events[i:])
```
```suggestion
			packet, err := ParsePacketFromEvents(events[i:i+1])
```
Is there any reason why we are not doing this?

crodriguezvega replied:
True, because it you use `packet, err := ParsePacketFromEvents(events[i:])` and you have a send packet event in index, let's say 4, then you would parse the same packet multiple times, right?

tbruyelle replied:
No because of the condition just above. Basically here, `events[i]` is always of type `EventTypeSendPacket`. Anyway I resolved this conversation because the code is outdated: the logic has been moved to an other function.

srdtrk commented on testing/events.go:
```diff
@@ -120,6 +120,22 @@ func ParsePacketFromEvents(events []abci.Event) (channeltypes.Packet, error) {
 
+// ParsePacketsFromEvents parses events emitted from a MsgRecvPacket and returns
+// all the packets found.
+func ParsePacketsFromEvents(events []abci.Event) ([]channeltypes.Packet, error) {
```
I'd actually prefer if the main logic was happening in this function and have `ParsePacketFromEvents` would call into this and pick the first element from the result. I'm ok either way though.

crodriguezvega replied:
Yeah, the implementation might become easier to reason like that.

tbruyelle replied:
Good idea, so I added a test and then moved the logic from `ParsePacketFromEvents` to `ParsePacketsFromEvents`.

srdtrk commented:
Thanks for this :)

crodriguezvega commented on testing/events.go:
```diff
@@ -120,6 +120,22 @@ func ParsePacketFromEvents(events []abci.Event) (channeltypes.Packet, error) {
+func ParsePacketsFromEvents(events []abci.Event) ([]channeltypes.Packet, error) {
+	var packets []channeltypes.Packet
+	for i, ev := range events {
+		if ev.Type == channeltypes.EventTypeSendPacket {
```
I think we could also remove this `if` statement, because the same one is in `ParsePacketFromEvents`.

tbruyelle replied:
done with the logic moved from one function to the other.

crodriguezvega reviewed (commented):
Thank you, @tbruyelle.

crodriguezvega commented:
@tbruyelle if you agree with the review comments, would you be able to work on them? Otherwise, if you're busy, we are happy to take over the PR and address the comments. If you give me permissions to push to your branch, I am happy to do that.

tbruyelle commented:
> @tbruyelle if you agree with the review comments, would you be able to work on them? Otherwise, if you're busy, we are happy to take over the PR and address the comments. If you give me permissions to push to your branch, I am happy to do that.

@crodriguezvega oups indeed I was super busy the past weeks, but today is quieter, I'll address the comments, thanks for the ping!

tbruyelle commented:
@crodriguezvega I addressed all the comments. As requested I moved the logic from one function to the other, and also added a test before that to ensure I didn't break anything during the process.

crodriguezvega commented on testing/events_test.go:
```diff
@@ -0,0 +1,223 @@
+)
+
+func TestParsePacketsFromEvents(t *testing.T) {
+	tests := []struct {
```
nit: this is the naming we use in the majority of other tests.

```suggestion
	testCases := []struct {
```

crodriguezvega commented on testing/events_test.go:
```diff
@@ -0,0 +1,223 @@
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+			assert := assert.New(t)
+
+			allPackets, err := ibctesting.ParsePacketsFromEvents(tt.events)
+
+			if tt.expectedError != "" {
+				assert.ErrorContains(err, tt.expectedError)
+			} else {
+				require.NoError(err)
+				assert.Equal(tt.expectedPackets, allPackets)
+			}
+
+			firstPacket, err := ibctesting.ParsePacketFromEvents(tt.events)
+
+			if tt.expectedError != "" {
+				assert.ErrorContains(err, tt.expectedError)
+			} else {
+				require.NoError(err)
+				assert.Equal(tt.expectedPackets[0], firstPacket)
+			}
+		})
+	}
```
I think this could written like this (assuming that `tests` is renamed to `testCases`):


```golang
for _, tc := range testCases {
  t.Run(tc.name, func(t *testing.T) {
    allPackets, err := ibctesting.ParsePacketsFromEvents(tc.events)

    if tc.expectedError == "" {
      require.NoError(t, err)
      require.Equal(t, tc.expectedPackets, allPackets)
    } else {
      require.ErrorContains(t, err, tc.expectedError)
    }

    firstPacket, err := ibctesting.ParsePacketFromEvents(tc.events)

    if tc.expectedError == "" {
      require.NoError(t, err)
      require.Equal(t, tc.expectedPackets[0], firstPacket)
    } else {
      require.ErrorContains(t, err, tc.expectedError)
    }
   })
}
```

So that:
- We don't need to import `assert`.
- We execute first the checks for the success case.

crodriguezvega commented on testing/events_test.go:
```diff
@@ -0,0 +1,223 @@
+			expectedError: "ibctesting.ParsePacketsFromEvents: strconv.ParseUint: parsing \"x\": invalid syntax",
+		},
+		{
+			name: "ok: events with packets",
```
We usually have the success case the first on of the slice.

```suggestion
			name: "success",
```

crodriguezvega commented on testing/events_test.go:
```diff
@@ -0,0 +1,223 @@
+		{
+			name:          "fail: no events",
+			events:        []abci.Event{},
+			expectedError: "ibctesting.ParsePacketsFromEvents: acknowledgement event attribute not found",
```
Since we are using `ErrorContains` it could be enough to just have this (same for the other failure test cases):

```suggestion
			expectedError: "acknowledgement event attribute not found",
```

crodriguezvega reviewed (commented):
Thank you for addressing the comments and adding a test, @tbruyelle. I have left a few more comments now. :)

tbruyelle commented on testing/events.go:
```diff
@@ -62,12 +62,28 @@ func ParseChannelIDFromEvents(events []abci.Event) (string, error) {
+// Returns an error if no packet is found.
+func ParsePacketsFromEvents(events []abci.Event) ([]channeltypes.Packet, error) {
+	ferr := func(err error) ([]channeltypes.Packet, error) {
+		return nil, fmt.Errorf("ibctesting.ParsePacketsFromEvents: %w", err)
```
As far as I know, the stdlib error package doesn't embed the callstack (unlike other more sophisticated error packages, for example).

This is an attempt to make it easier to find the source of the error. But I can understand if this doesn't suit the code style of the repo. Let me know if you want me to just return unlabelled errors.

tbruyelle commented:
@crodriguezvega comments addressed here https://github.com/cosmos/ibc-go/pull/4975/commits/8e06417a36c5eb0b78bbaba42e0ba27cfc6253b5

- rename tests to testCases, tt to tc
- remove assert package usage
- revert `if tc.expectedError` condition
- rename success case to "success", move it on top
- remove `expectedError` common prefix

crodriguezvega reviewed (approved):
Nice work, @tbruyelle. Thank you for addressing all the comments!

If you rebase with main, I will try to merge this before your branch gets out of sync again. 🤞

tbruyelle commented:
> Nice work, @tbruyelle. Thank you for addressing all the comments!
> 
> If you rebase with main, I will try to merge this before your branch gets out of sync again. 🤞

Thank you Carlos, I've just rebased on main!

chatton commented:
ah, those 2 failures are not real failures, feel free to disregard them in this PR
